### ISSUE TITLE
mini-batch is very slow and takes a lot of memory

### ISSUE THREAD

**Issue by qzhu2017** [qzhu2017/PyXtal_FF#15]
@David-Zagaceta @yanxon 
Here is another unpleasant issue. In the following code,
https://github.com/qzhu2017/PyXtal_FF/blob/master/examples/SNU-SiO2/so4-mini.py

I am trying to use the minibatch for the training of 10000 SiO2 structures. Even I specified 200 structures for each batch (with the option of `memory: out`), the code still consumes a lot of memory. It also makes the code very slow. Have you guys ever ran into this issue?

**Comment by qzhu2017** [comment]
This will be fixed later after v-0.1. A possible solution is to split the entire database into many .npy files based on the given batch size. Then we either pick each batch for training.

**Comment by yanxon** [comment]
@qzhu2017 

It is slow because of the `shelve` database. It takes a lot of memory because the `forces` are store in the memory while the `x`, `dxdr`, and `rdxdr` are in memory.

**Comment by qzhu2017** [comment]
@yanxon That’s why we will use a set of npy file at the beginning. And then use these npy files for the subsequent trainings. We discussed this some time ago, didn’t we?

**Comment by yanxon** [comment]
@qzhu2017 

Yes. I'm working on this npy files right now. I just want to record what's wrong with the previous version for future reference (if needed).

I'm not quite done yet. But, here is a minor issue:
The npy file requires all arrays to be the shape. Else, we will encounter slow loading (just like shelve). To remedy this, I have to create the same size for all arrays with respect to `max atoms`.

**Comment by qzhu2017** [comment]
@yanxon If that is the case, can you just resort the structures by the number of atoms? When you split them into different batches, make sure that the structures with the similar sizes are assigned to the same batch.

**Comment by yanxon** [comment]
@qzhu2017 @David-Zagaceta 

I truly believe the memory-consuming part about mini-batch comes from the Initial state:
https://github.com/qzhu2017/PyXtal_FF/blob/95c1e2812124324903aa4dc1df82fc680805984b/pyxtal_ff/models/neuralnetwork.py#L235-L248

The Initial state basically tells you the current MAE values with no parameters update. I did this to ensure that if you want to pick up the training for more training, the MAE values from previous iteration should be equal to the Initial state MAEs. 

As @qzhu2017 and I discussed about it last night, PyTorch actually didn't release the `x`, `dxdr`, etc. in the initial state just in case if users would like to update the parameters later. However, if you run the true mini-batch operation as shown here:
https://github.com/qzhu2017/PyXtal_FF/blob/95c1e2812124324903aa4dc1df82fc680805984b/pyxtal_ff/models/neuralnetwork.py#L253-L257

Here, PyTorch will release the `x`, `dxdr`, etc. after every parameters update.

**Comment by qzhu2017** [comment]
@yanxon Any updates here?

**Comment by yanxon** [comment]
@qzhu2017 

It seems like NPY is troublesome if you want to scale the calculation.

Here is what happened to me:
First, I ran SiO2 calculation with NPY with batch_size of 50. According to my rough counting of time, each calculate loss would take ~2-3 seconds. Then, I ran with shelve in parallel with exactly the same setup to measure the time. The calculation with NPY for each calculate loss took ~4-5 seconds.

Then I truly run the both NPY and Shelve in parallel to repeat **Shelve / NPY** with `10,000 SiO2` data set (adam, memory out) -- batch_size=50. The results are very different:
peak memory usage: / ~5%
training time (5 epochs): 3029.89s / 5323s

**Comment by qzhu2017** [comment]
@yanxon 
Can you just decrease the number of npy files?  As I said, we don't really need that many npy files. 

Please be efficient. I don't know why this takes such a long time.

**Comment by yanxon** [comment]
@qzhu2017 

New benchmark:
```
Shelve / NPY with 10,000 SiO2 data set (adam, memory out) -- batch_size=50
peak memory usage: / ~5%
training time (5 epochs): 3029.89s / 1772.20s

Shelve / NPY with 10,000 SiO2 data set (adam, memory out) -- batch_size=25
peak memory usage: / ~3%
training time (5 epochs): / 1810.82s

Shelve / NPY with 10,000 SiO2 data set (adam, memory out) -- batch_size=200
peak memory usage: / ~13%
training time (5 epochs): / 1818.09s

Shelve / NPY with 10,000 SiO2 data set (adam, memory out) -- Increasing bundle to 1000; batch_size = 50
peak memory usage: / ~5%
training time (5 epochs): / 1350.90s

Shelve / NPY with 10,000 SiO2 data set (adam, memory out) -- Increasing bundle to 1000; batch_size = 200
peak memory usage: / ~12%
training time (5 epochs): / 1339.01

Bundle 1000 less files to be open, Only expensive is in Iteration 1

For each iteration:
Iteration 1: 265.64 s
Iteration 2: 266.22 s
Iteration 3: 265.96 s
Iteration 4: same
Iteration 5: same
```

**Comment by yanxon** [comment]
@qzhu2017 

I mentioned that the NN code will open all of the NPY files only once. The timing is even worse than using `deepcopy`. As batch_size increases, the calculation becomes slower. I don't exactly know why. The einsum took the most effort here, which I don't know exactly why.
```
 445 The training time: 371.82 s
 446 The Neural Network Potential is exported to SiO2-mini-batch/30-30-checkpoint.pth
 447
 448
 449 ==================================================================
 450
 451 ==================== Evaluating Training Set ====================
 452
 453          6246763 function calls (5929933 primitive calls) in 376.118 seconds
 454
 455    Ordered by: internal time
 456
 457    ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 458     20000  275.525    0.014  275.525    0.014 {built-in method einsum}
 459     20200   32.449    0.002   32.449    0.002 {method 'run_backward' of 'torch._C._EngineBase' objects}
 460     60000   27.879    0.000   27.879    0.000 {built-in method addmm}
 461       200   15.324    0.077  354.508    1.773 neuralnetwork.py:466(calculate_loss)
 462     40000    2.144    0.000    2.144    0.000 {built-in method tanh}
 463     20000    1.783    0.000    1.783    0.000 {built-in method torch._C._nn.l1_loss}
 464    330000    1.585    0.000    3.447    0.000 memmap.py:330(__getitem__)
 465      1060    1.431    0.001    1.431    0.001 {method 'read' of '_io.FileIO' objects}
 466 120000/20000    1.275    0.000   36.399    0.002 module.py:540(__call__)
 467     60000    1.158    0.000    1.158    0.000 {method 't' of 'torch._C._TensorBase' objects}
 468         1    1.078    1.078  372.154  372.154 neuralnetwork.py:164(train)
 469     30000    1.033    0.000    1.033    0.000 {method 'sum' of 'torch._C._TensorBase' objects}

```

**Comment by qzhu2017** [comment]
@yanxon Just to be clear, I suggest the following
1, Do not create too many npy files. Ideally, Let’s just keep the following
- Atom based properties, forces, atom types
- structure based properties, energy
- dxdr
- rdxdr 
I think we just need 4 npy files for each bundle

2, Get rid of the data loader. Say if we have 10000 structures and the bundle is 1000, we will have 10 bundles. When the batch size is 100, we will have 100 batches. Let’s create a list from 1 to 100. The correspond to 
- List 1-10 -> 01.npy
- List 11-20 -> 02.npy
...

For each epoch, we generate a randomly permutated list [0, 2, 100, 4, .....], then pick the structures based on this list. This way, you only need to open 1-2 npy files to for each mini batch.

There is no need to make the code over complicated.

**Comment by qzhu2017** [comment]
@yanxon Actually, I think your memmap version is not bad. It seems to be somehow random. However, the average time for each epoch is about 80-90 seconds if you continue the simulation.

**Comment by yanxon** [comment]
@qzhu2017 

Shall we just proceed to memmap? I will finish the entire code with memmap.

**Comment by qzhu2017** [comment]
@yanxon Hold on. I am running some benchmarks now. It seems that the shelve version is comparable to memmap. I tried to continue the training. Sometimes, it seems to be even faster.

On cms, 
## 10,000 SiO2 data set (adam, memory out) 
### Shelve 
peak memory usage: / ~5%
training time (5 epochs): 846s/630s/395s

### memmap 
peak memory usage: / ~1%
training time (5 epochs): 600s/450s/455s

Change the batch size does not quite affect the time.
